bindings/python/ethash/__init__.py

```diff
@@ -1,21 +1,39 @@
 # ethash: C/C++ implementation of Ethash, the Ethereum Proof of Work algorithm.
 # Copyright 2019 Pawel Bylica.
+# Copyright 2023 Sam Wilson.
 # Licensed under the Apache License, Version 2.0.
 
-from _ethash import ffi, lib
+from _ethash import ffi, lib  # type: ignore
 
+from typing import TYPE_CHECKING, Tuple
 
-def keccak_256(data):
+from collections.abc import Sized
+
+if TYPE_CHECKING:
+    from typing import Protocol
+
+    from _typeshed import WriteableBuffer, ReadableBuffer
+
+    class SizedWriteableBuffer(Protocol, WriteableBuffer, Sized):
+        pass
+
+    class SizedReadableBuffer(Protocol, ReadableBuffer, Sized):
+        pass
+
+
+def keccak_256(data: "SizedWriteableBuffer") -> bytes:
```

Comment:
I think that makes total sense! Since the default value of `require_writable` is `False`, I've omitted it and change this to use `SizedReadableBuffer` throughout.

Reply:
Since readable *and* writeable buffer types work for the `FFI.from_buffer()` method, I think using a class defined similarly to `SizedBuffer(Protocol, ReadableBuffer, WriteableBuffer, Sized)` would work well.

Reply:
I'm going off the definitions from typeshed:

https://github.com/python/typeshed/blob/32535ee8952dffecbe7a9fd04e5ed8ba0f1c8da2/stubs/cffi/cffi/api.pyi#L49-L52

https://github.com/python/typeshed/blob/32535ee8952dffecbe7a9fd04e5ed8ba0f1c8da2/stdlib/_typeshed/__init__.pyi#L247

It looks like `ReadableBuffer` is a super-type of `WritableBuffer`.

Reply:
Ah, and that makes perfect sense 👌🏼

Reply:
On the C side we want `require_writable=False` because we never write to it.